aten/src/ATen/cuda/LegacyTHFunctionsCUDA.cpp

```diff
@@ -171,136 +171,47 @@ std::tuple<Tensor,Tensor> _th_sort_stable(const Tensor & self, c10::optional<boo
             break;
         }
         case ScalarType::Double: {
-            auto self_ = checked_dense_tensor_unwrap(self, "self", 1, "_th_topk", false, DeviceType::CUDA, dispatch_scalar_type);
-            THCudaDoubleTensor_topk(globalContext().getTHCState(), values_, indices_, self_, k, dim, largest, sorted);
+            THCudaDoubleTensor_sortKeyValueInplace(globalContext().getTHCState(), keys_, values_, dim, descending);
             break;
         }
         case ScalarType::Float: {
-            auto self_ = checked_dense_tensor_unwrap(self, "self", 1, "_th_topk", false, DeviceType::CUDA, dispatch_scalar_type);
-            THCudaTensor_topk(globalContext().getTHCState(), values_, indices_, self_, k, dim, largest, sorted);
+            THCudaTensor_sortKeyValueInplace(globalContext().getTHCState(), keys_, values_, dim, descending);
             break;
         }
         case ScalarType::Int: {
-            auto self_ = checked_dense_tensor_unwrap(self, "self", 1, "_th_topk", false, DeviceType::CUDA, dispatch_scalar_type);
-            THCudaIntTensor_topk(globalContext().getTHCState(), values_, indices_, self_, k, dim, largest, sorted);
+            THCudaIntTensor_sortKeyValueInplace(globalContext().getTHCState(), keys_, values_, dim, descending);
             break;
         }
         case ScalarType::Long: {
-            auto self_ = checked_dense_tensor_unwrap(self, "self", 1, "_th_topk", false, DeviceType::CUDA, dispatch_scalar_type);
-            THCudaLongTensor_topk(globalContext().getTHCState(), values_, indices_, self_, k, dim, largest, sorted);
+            THCudaLongTensor_sortKeyValueInplace(globalContext().getTHCState(), keys_, values_, dim, descending);
             break;
         }
         case ScalarType::Short: {
-            auto self_ = checked_dense_tensor_unwrap(self, "self", 1, "_th_topk", false, DeviceType::CUDA, dispatch_scalar_type);
-            THCudaShortTensor_topk(globalContext().getTHCState(), values_, indices_, self_, k, dim, largest, sorted);
+            THCudaShortTensor_sortKeyValueInplace(globalContext().getTHCState(), keys_, values_, dim, descending);
             break;
         }
         case ScalarType::Half: {
-            auto self_ = checked_dense_tensor_unwrap(self, "self", 1, "_th_topk", false, DeviceType::CUDA, dispatch_scalar_type);
-            THCudaHalfTensor_topk(globalContext().getTHCState(), values_, indices_, self_, k, dim, largest, sorted);
-            break;
-        }
-        case ScalarType::BFloat16: {
-            auto self_ = checked_dense_tensor_unwrap(self, "self", 1, "_th_topk", false, DeviceType::CUDA, dispatch_scalar_type);
-            THCudaBFloat16Tensor_topk(globalContext().getTHCState(), values_, indices_, self_, k, dim, largest, sorted);
+            THCudaHalfTensor_sortKeyValueInplace(globalContext().getTHCState(), keys_, values_, dim, descending);
             break;
         }
         default:
```

Comment:
I think for this you need to add a BFloat16 branch, this brings back the existing partial bfloat16 support. But bfloat16 is still not fully supported due to sort, for example:
```python
>>> import torch
>>> a = torch.randn(2, 4096, device='cuda', dtype=torch.bfloat16)
>>> a.topk(4000)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: "sort" not implemented for 'BFloat16'
Exception raised from operator() at /home/gaoxiang/pytorch-master/aten/src/ATen/native/cuda/Sort.cu:162 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x5c (0x7f02da9eed8c in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x64 (0x7f02da9c19d6 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libc10.so)
frame #2: at::native::sort_out_stable_cuda(at::Tensor const&, c10::optional<bool>, long, bool, at::Tensor&, at::Tensor&) + 0x1868 (0x7f02dc5b0488 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cuda.so)
frame #3: at::native::sort_out_cuda(at::Tensor const&, long, bool, at::Tensor&, at::Tensor&) + 0x9f (0x7f02dc5b05ef in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cuda.so)
frame #4: THCudaBFloat16Tensor_topk(THCState*, c10::TensorImpl*, c10::TensorImpl*, c10::TensorImpl*, long, int, int, int) + 0x472 (0x7f02db5bdb12 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cuda.so)
frame #5: <unknown function> + 0x25d124c (0x7f02dcfe424c in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cuda.so)
frame #6: <unknown function> + 0x262498e (0x7f02dd03798e in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cuda.so)
frame #7: <unknown function> + 0x2624a4b (0x7f02dd037a4b in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cuda.so)
frame #8: at::topk_out(at::Tensor&, at::Tensor&, at::Tensor const&, long, long, bool, bool) + 0x1aa (0x7f030a65349a in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #9: at::native::topk(at::Tensor const&, long, long, bool, bool) + 0xcf (0x7f030a24892f in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #10: <unknown function> + 0x2624bcb (0x7f02dd037bcb in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cuda.so)
frame #11: <unknown function> + 0x2624c83 (0x7f02dd037c83 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cuda.so)
frame #12: at::redispatch::topk(c10::DispatchKeySet, at::Tensor const&, long, long, bool, bool) + 0xdd (0x7f030a8a271d in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #13: <unknown function> + 0x2fb95ad (0x7f030c1125ad in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #14: <unknown function> + 0x2fb9836 (0x7f030c112836 in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #15: at::Tensor::topk(long, long, bool, bool) const + 0x15f (0x7f030ad49bef in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so)
frame #16: <unknown function> + 0x652faa (0x7f0312760faa in /home/gaoxiang/.local/lib/python3.9/site-packages/torch/lib/libtorch_python.so)
<omitting python frames>
frame #29: __libc_start_main + 0xd5 (0x7f033fb9db25 in /usr/lib/libc.so.6)


>>> b = torch.randn(2, 1024, device='cuda', dtype=torch.bfloat16)
>>> b.topk(10)
torch.return_types.topk(
values=tensor([[3.3906, 3.2812, 2.9062, 2.7812, 2.7812, 2.5000, 2.4219, 2.3906, 2.3281,
         2.3125],
        [4.0938, 3.5938, 3.1406, 2.8906, 2.8594, 2.7969, 2.7812, 2.7344, 2.6719,
         2.5156]], device='cuda:0', dtype=torch.bfloat16),
indices=tensor([[787, 919, 909, 803, 499, 610, 492,   5, 456, 197],
        [394, 630, 226, 869, 189, 156, 757, 808, 353, 778]], device='cuda:0'))
```
Could you open an issue for the above issue?